File: 3.x.x/crates-utopia/pools/src/adapter.rs
//! PHP `Utopia\Pools\Adapter`.

use std::time::Duration;

use async_trait::async_trait;

use crate::{Connection, PoolError};

pub mod stack;
pub mod swoole;

pub use stack::Stack;
pub use swoole::Swoole;

/// Storage and synchronisation for a pool's idle resources.
///
/// The pool owns capacity accounting; an adapter only holds what is idle and
/// serialises access to the pool's bookkeeping.
#[async_trait]
pub trait Adapter<T: Send>: Send + Sync {
    /// Prepare to hold up to `size` idle resources. A sizing hint, not a cap.
    fn initialize(&self, size: usize);

    /// Push an idle connection.
    fn push(&self, connection: Connection<T>);

    /// Take an idle resource, waiting up to `timeout` for one to arrive.
    ///
    /// [`Stack`] ignores `timeout` and returns immediately.
    async fn pop(&self, timeout: Duration) -> Option<Connection<T>>;

    /// Number of idle connections currently held.
    fn count(&self) -> usize;

    /// Run `callback` atomically with respect to other pool operations.
    ///
    /// [`Stack`] has no concurrency and runs the callback directly.
    fn synchronized(&self, callback: Box<dyn FnOnce() + Send>) -> Result<(), PoolError>;
}

File: src/Pools/Adapter.php
<?php

declare(strict_types=1);

namespace Utopia\Pools;

/**
 * Storage and synchronisation for a pool's idle resources.
 *
 * The pool owns capacity accounting; an adapter only holds what is idle and
 * serialises access to the pool's bookkeeping.
 */
abstract class Adapter
{
    /**
     * Prepare to hold up to $size idle resources. The pool enforces the limit,
     * so $size is a sizing hint rather than a cap the adapter must police.
     */
    abstract public function initialize(int $size): static;

    abstract public function push(mixed $connection): static;

    /**
     * Take an idle resource, waiting up to $timeout seconds for one to arrive.
     *
     * Adapters without concurrency return immediately: with a single execution
     * context nothing can hand a resource back while the caller waits, so
     * waiting cannot change the answer.
     *
     * @return mixed The resource, or null/false when none is available.
     */
    abstract public function pop(float $timeout): mixed;

    abstract public function count(): int;

    /**
     * Release anyone currently blocked in {@see pop()}, without handing over a
     * resource. They will find out for themselves whether the state they were
     * waiting on has changed; this says only that it is worth looking again.
     *
     * Adapters with no way to block satisfy this by construction: a caller that
     * never waits has nothing to release.
     */
    public function unblock(): static
    {
        return $this;
    }

    /**
     * Run $callback atomically with respect to other pool operations.
     *
     * Adapters without concurrency satisfy this by construction and may call
     * $callback directly.
     *
     * @return mixed The callback's return value.
     */
    abstract public function synchronized(callable $callback): mixed;
}
